### PR TITLE
Remove n+1

### PR DESCRIPTION
**Pull request by daniely**
some db optimizations mentioned in #386

most of the legwork was done by @phlipper. i simply did some cleanup and applied it to the newest master branch.

i'm splitting up the n+1 and counter cache into separate commits

### CODE REVIEW

File: app/controllers/forem/forums_controller.rb [hunk 1]
```diff
@@ -4,7 +4,7 @@ class ForumsController < Forem::ApplicationController
     helper 'forem/topics'
 
     def index
-      @categories = Forem::Category.all
+      @categories = Forem::Category.includes(:forums => [:category, :moderators]).all
```
Comment (parndt) [inline app/controllers/forem/forums_controller.rb]: This means we don't need the `.all` anymore

File: app/controllers/forem/categories_controller.rb [hunk 2]
```diff
@@ -1,6 +1,14 @@
...
   class CategoriesController < Forem::ApplicationController
+    before_filter :find_category, :only => :show
+
     helper 'forem/forums'
     load_and_authorize_resource :class => 'Forem::Category'
+
+    private
+
+    def find_category
+      @category = Forem::Category.includes(:forums => :category).find(params[:id])
```
Comment (radar) [inline app/controllers/forem/categories_controller.rb:11]: I don't understand the `includes` here. Why is that necessary if all the forums have the same category? Shouldn't it just be `includes(:forums)`?

File: app/controllers/forem/forums_controller.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@ class ForumsController < Forem::ApplicationController
     helper 'forem/topics'
 
     def index
-      @categories = Forem::Category.all
+      @categories = Forem::Category.includes(:forums => [:category, :moderators])
```
Comment (radar) [inline app/controllers/forem/forums_controller.rb:7]: I don't think the `category` should be here in the includes, either.

File: app/controllers/forem/forums_controller.rb [hunk 4]
```diff
@@ -16,7 +16,8 @@ def show
         @forum.topics.visible.approved_or_pending_review_for(forem_user)
       end
 
-      @topics = @topics.by_pinned_or_most_recent_post.page(params[:page]).per(Forem.per_page)
+      @topics = @topics.by_pinned_or_most_recent_post.includes(:forum, :user)
```
Comment (radar) [inline app/controllers/forem/forums_controller.rb:19]: Why does `forum` need to be included here?

File: app/controllers/forem/topics_controller.rb [hunk 5]
```diff
@@ -92,7 +92,7 @@ def find_forum
     end
 
     def find_posts(topic)
-      posts = topic.posts
+      posts = topic.posts.includes(:user, :reply_to, :topic)
```
Comment (radar) [inline app/controllers/forem/topics_controller.rb:95]: Why does `topic` need to be included here?
